src/ucp/core/ucp_worker.c

```diff
@@ -2609,35 +2609,53 @@ ucp_worker_discard_uct_ep_purge(uct_pending_req_t *self, void *arg)
     ucp_worker_discard_uct_ep_complete(req);
 }
 
-static void ucp_worker_discard_uct_ep_cleanup(ucp_worker_h worker)
+static int ucp_worker_discard_uct_ep_hash_retrieve_one(ucp_worker_h worker,
+                                                       ucp_request_t **req_p)
 {
     ucp_request_t *req;
-    ucp_ep_h ucp_ep;
     uct_ep_h uct_ep;
-
-    /* Destroying UCP EP in ucp_ep_disconnected() could start UCT EP discarding
-     * operations. Do cleanup of discarding functionality after trying to
-     * destroy UCP EPs in order to destroy all remaining UCP EPs here (they are
-     * destroyed during discarding operation completion for all UCT EPs) */
+    khiter_t iter;
 
     kh_foreach(&worker->discard_uct_ep_hash, uct_ep, req, {
         ucs_assert(uct_ep == req->send.discard_uct_ep.uct_ep);
 
+        iter = kh_get(ucp_worker_discard_uct_ep_hash,
+                      &worker->discard_uct_ep_hash, uct_ep);
+        ucs_assert(kh_value(&worker->discard_uct_ep_hash, iter) == req);
+        kh_del(ucp_worker_discard_uct_ep_hash, &worker->discard_uct_ep_hash,
+               iter);
+
+        *req_p = req;
+        return 1;
+    })
+
+    return 0;
+}
+
+static void ucp_worker_discard_uct_ep_cleanup(ucp_worker_h worker)
+{
+    ucp_request_t *req;
+    ucp_ep_h ucp_ep;
+    uct_ep_h uct_ep;
+
+    while (ucp_worker_discard_uct_ep_hash_retrieve_one(worker, &req)) {
```

Comment:
why need to loop?

Reply:
to remove UCT EPs from discarding hash for which the discard done,
so, `ucp_worker_discard_uct_ep_cleanup` could be called twice then

Reply:
maybe move hash to temp list?

Reply:
then we anyway need to do `kh_get(uct_ep)` to get an iterator `iter` and then do `kh_del(iter)` for all elements when going over a temp list.
also, we will have two loops:
- over khash to collect all discarding requests and put them to a temp list
- over the temp list to complete discarding operations and delete all elements

this approach smees more complicated for me.